contracts/DebtSettler.sol

```diff
@@ -56,73 +73,105 @@ contract DebtSettler is ReentrancyGuard {
 
-        for (uint i = 0; i < _length; i++) {
+        for (i; i < _length; i++) {
             address _borrower = _borrowers.at(i);
-            uint _credit = _amount * _debts[i] / _totalDebt;
-            (, uint _currentCredit) = _records.tryGet(_borrower);
+            uint _credit = _amount * _debt(_borrower) / _totalDebt;
+            (, uint _currentCredit) = _usersCredit.tryGet(_borrower);
 
             // we have to accumulate each time =)
-            _records.set(_borrower, _credit + _currentCredit);
+            _usersCredit.set(_borrower, _credit + _currentCredit);
+
+            // each loop use aprox 80k of gas
+            if (gasleft() <= 100_000) {
+                _lastIndexBuilt = i;
+                return;
+            }
         }
+
+        _lastIndexBuilt = 0; // ensure that if ends always starts from 0
     }
 
     /**
      * @dev Settle the debt to all borrowers registered using the build() function.
      *
      * If for some reason it runs out of gas, it will stop on the last borrower that could be settled.
      */
-    function pay() external nonReentrant {
-        asset.approve(address(pool), asset.balanceOf(address(this)));
+    function pay() external onlyHandler nonReentrant {
+        // Ensure always pay after build is finished
+        if (_lastIndexBuilt > 0) revert StillBuilding();
+
+        asset.approve(address(pool), _lastCredit);
 
-        for (uint i = 0; i < _records.length(); i++) {
-            (address _borrower, uint _debt) = _records.at(i);
+        // keep going from last paid
+        uint i = _lastIndexPaid == 0 ? 0 : (_lastIndexPaid + 1);
 
+        // just in case the records decrease in size before pay
+        if (i > _usersCredit.length()) i = _lastIndexPaid = 0;
+
+        for (i; i < _usersCredit.length(); i++) {
             // We should check for gasleft here, so we can repay the rest in the next tx if needed
-            if (gasleft() <= 50_000) break;
+            (address _borrower, uint _credit) = _usersCredit.at(i);
 
-            if (_debt > 0) {
-                pool.repayFor(_borrower, _debt);
-                _records.set(_borrower, 0);
+            if (_credit > 0 && dToken.balanceOf(_borrower) > 0) {
+                pool.repayFor(_borrower, _credit);
+                _usersCredit.set(_borrower, 0);
             }
-        }
+
+            // each loop use aprox 50k of gas
+            if (gasleft() <= 70_000) {
+                _lastIndexPaid = i;
+                return;
+            }
+        } // for
+
+        _lastIndexPaid = 0; // ensure that if ends always starts from 0
     }
 
     /**
      * @dev Clean up the list of borrowers which have no debt left.
      */
-    function clean() external nonReentrant {
-        address[] memory _toRemove = new address[](_records.length());
+    function clean() external onlyHandler nonReentrant {
+        address[] memory _toRemove = new address[](_usersCredit.length());
```

Comment:
```suggestion
        if (_lastIndexPaid > 0) revert StillPaying();
        address[] memory _toRemove = new address[](_usersCredit.length());
```
Just in case =)